### ISSUE TITLE
Probable miscompilation involving complex bit manipulation code and a specific subfunction coincidence

### ISSUE THREAD

**Issue by EDI-Systems** [cc65/cc65#2461]
The compiler does not seem to handle complex bitfield manipulations when it coincides with parameter clobbering.
See  [this godbolt snippet](https://godbolt.org/z/8rhz1PboP) for the minimal offending example.
This seems to be a bug in cc65, but is unclear. Other compilers (gcc,armcc,etc) work fine with the code.

**Comment by mrdudz** [comment]
Does it work when you compile without optimizations?

**Comment by EDI-Systems** [comment]
It seems to work in this small example, but don't work in my project.

**Comment by mrdudz** [comment]
In that case it'd be helpful to have a minimal test that reproduces the problem when optimizations are disabled

**Comment by EDI-Systems** [comment]
Maybe I remembered wrongly. It seemed to work without optimizations.

**Comment by EDI-Systems** [comment]
I can't reproduce it anymore without optimizations, and maybe I forgot to recompile.

**Comment by EDI-Systems** [comment]
Hello, can the developers confirm that this is a bug or I misreported it as a bug?

**Comment by mrdudz** [comment]
Sorry, i'll have a look when i find some time to do it at the weekend.

**Comment by EDI-Systems** [comment]
> Sorry, i'll have a look when i find some time to do it at the weekend.

Thank you very much for your time into this matter! I'm looking forward to your reply, and will also have some time to help you look into this over the weekend.

**Comment by mrdudz** [comment]
What would help: try selectively disabling some optimizer steps (see docs) - with some luck you can track down the step that produces the problem

**Comment by EDI-Systems** [comment]
> What would help: try selectively disabling some optimizer steps (see docs) - with some luck you can track down the step that produces the problem

Even a single -O will trigger the issue. Any of -Oi, -Os, -Or or their comb will also trigger the peoblem.

**Comment by mrdudz** [comment]
That is not what i meant - you can disable specific optimizer steps using --disable-opt

**Comment by mrdudz** [comment]
Ok first conclusion: i can reproduce it, and can confirm it is a bug :) It's also tied to optimizations, which means we might be able to track down (and fix) the optimizer step like i said.

**Comment by EDI-Systems** [comment]
OptStackOps This is a offender.

**Comment by EDI-Systems** [comment]
OptUnusedLoads could also cause changes in code but it seems that it's just doing benigh transformations.

**Comment by EDI-Systems** [comment]
https://github.com/EDI-Systems/M5P01_Prokaron/blob/730520dc3935f78c81543e762b6c48cb7ba910df/Source/Kernel/rmp_kernel.c#L3679

This is where the original project's location where it produced the offending code. 
Later I realized that (num&(~WORD_MASK))+(num&WORD_MASK) = num, and rewritten the code to run more efficiently and the issue also went away.

**Comment by mrdudz** [comment]
Made a testprog that will work in the testbench from your example (unfortunately cant drop it into "todo" now, since it actually works without optimizations)

[opt-bug-2463.zip](https://github.com/user-attachments/files/16116600/opt-bug-2463.zip)

One further conclusion so far: it behaves the same with 6502 and 65C02

**Comment by mrdudz** [comment]
> OptStackOps This is a offender.

yup. now we need to stare at it and find out whats wrong
```
$ cl65 -Osir -Wc --disable-opt -Wc OptStackOps -t sim6502 -o bla.bin opt-bug-2463.c && sim65 bla.bin
fixed: 0x4715 original: 0x4715 wrong: 0x4715
fixed: 0x811 original: 0x811 wrong: 0x811
failures: 0
```

**Comment by mrdudz** [comment]
If anyone wants to stare at it as well, its here:

https://github.com/cc65/cc65/blob/0541b65aa48ea2508b218ff7097911e741b4bc12/src/cc65/coptstop.c#L1841

**Comment by mrdudz** [comment]
@EDI-Systems could you try with #2502 applied and tell if it makes your project work?

**Comment by EDI-Systems** [comment]
> @EDI-Systems could you try with #2502 applied and tell if it makes your project work?

I would try when I have time, but these days I've been pretty busy..
I think I would have time into this before the next weekend.

**Comment by EDI-Systems** [comment]
I'll try to find a little bit of time tonight. Hopefully that's gonna test it out.

**Comment by EDI-Systems** [comment]
On my side stuff looks like fixed, missing instruction is back, and old benchmark working again

Before the fix:
![before](https://github.com/user-attachments/assets/393b41af-8553-4e6a-8f11-54c5e57e7cf9)

After the fix:
![after](https://github.com/user-attachments/assets/41d9238d-bf7a-4b0a-889b-699ba7d5072f)

**Comment by EDI-Systems** [comment]
If this also passes your tests, I think we can close the case.

**Comment by mrdudz** [comment]
there still seem to be problems, see #2942

**Comment by mrdudz** [comment]
closed via #2951
